subprojects/gst-plugins-bad/sys/msdk/gstmsdkallocator_libva.c

```diff
@@ -104,158 +91,62 @@ gst_msdk_frame_alloc (mfxHDL pthis, mfxFrameAllocRequest * req,
-#endif
-
-    va_status = vaCreateSurfaces (gst_msdk_context_get_handle (context),
-        format,
-        req->Info.Width, req->Info.Height, surfaces, surfaces_num, attribs,
-        num_attribs);
-
-    status = gst_msdk_get_mfx_status_from_va_status (va_status);
-    if (status != MFX_ERR_NONE) {
-      GST_WARNING ("failed to create VA surface");
-      return status;
-    }
+    gst_buffer_pool_set_config (pool, config);
+    gst_buffer_pool_set_active (pool, TRUE);
 
     for (i = 0; i < surfaces_num; i++) {
-      /* Get dmabuf handle if MFX_MEMTYPE_EXPORT_FRAME */
-      if (req->Type & MFX_MEMTYPE_EXPORT_FRAME) {
-        VADRMPRIMESurfaceDescriptor va_desc = { 0 };
-        uint32_t export_flags = VA_EXPORT_SURFACE_SEPARATE_LAYERS |
-            VA_EXPORT_SURFACE_READ_WRITE;
-
-        va_status =
-            vaExportSurfaceHandle (gst_msdk_context_get_handle (context),
-            surfaces[i], VA_SURFACE_ATTRIB_MEM_TYPE_DRM_PRIME_2, export_flags,
-            &va_desc);
-
-        status = gst_msdk_get_mfx_status_from_va_status (va_status);
-
-        if (MFX_ERR_NONE != status) {
-          GST_ERROR ("Failed to export surface");
-          return status;
-        }
-
-        g_assert (va_desc.num_objects);
-
-        /* This plugin supports single object only */
-        if (va_desc.num_objects > 1) {
-          GST_ERROR ("Can not support multiple objects");
-          return MFX_ERR_UNSUPPORTED;
-        }
-
-        msdk_mids[i].desc = va_desc;
-      }
-
-      /* Don't use image for DMABuf */
-      msdk_mids[i].image.image_id = VA_INVALID_ID;
-      msdk_mids[i].image.buf = VA_INVALID_ID;
+      GstBuffer *buf;
 
-      msdk_mids[i].surface = &surfaces[i];
-      mids[i] = (mfxMemId *) & msdk_mids[i];
-    }
-  } else {
-    /* This is requested from the driver when h265 encoding.
-     * These buffers will be used inside the driver and released by
-     * gst_msdk_frame_free functions. Application doesn't need to handle these buffers.
-     *
-     * See https://github.com/Intel-Media-SDK/samples/issues/13 for more details.
-     */
-    VAContextID context_id = req->AllocId;
-    gint width32 = 32 * ((req->Info.Width + 31) >> 5);
-    gint height32 = 32 * ((req->Info.Height + 31) >> 5);
-    guint64 codedbuf_size = (width32 * height32) * 400LL / (16 * 16);
-
-    for (i = 0; i < surfaces_num; i++) {
-      VABufferID coded_buf;
-
-      va_status = vaCreateBuffer (gst_msdk_context_get_handle (context),
-          context_id, VAEncCodedBufferType, codedbuf_size, 1, NULL, &coded_buf);
-
-      status = gst_msdk_get_mfx_status_from_va_status (va_status);
-      if (status < MFX_ERR_NONE) {
-        GST_ERROR ("failed to create buffer");
-        return status;
+      if (gst_buffer_pool_acquire_buffer (pool, &buf, NULL) != GST_FLOW_OK) {
+        GST_ERROR ("Failed to allocate buffer");
+        gst_buffer_pool_set_active (pool, FALSE);
+        gst_object_unref (pool);
+        return MFX_ERR_MEMORY_ALLOC;
```

Comment:
Do we need to free the malloced memory before return?

Reply:
Yes, fixed this. Thanks.